### DISCUSSION TITLE
Spherical coordinates and implicit (numerical) equilibria

### DISCUSSION THREAD

**Discussion by vreville** [n-claes/legolas#162] [Q&A] [answered]
Hi, 

I don't know if this question will reach anyone, but I was wondering whether it was possible to set up spherical equilibrium, and if so, if one could define the equilibrium numerically (when there is no analytical formulation)? 

Thanks, 

Victor

**Comment by n-claes** [comment, marked as answer]
Hi Victor, currently only Cartesian and cylindrical equilibria are supported, with 1D variation along $x$ and $r$, respectively. For detailed information on the various setups that are possible you can take a look at the various papers with applications listed on the website: https://dev.legolas.science/publications/.

To answer your second question first, you can always numerically define the equilibria. We did some applications with numerical equilibria in solar atmospheric conditions, where "equilibrium" was obtained by integrating an ODE obtained from force/thermal balance. As long as your sampling on the grid is sufficient to approach force/thermal balance it's perfectly fine.
Note that you can run the code for backgrounds that are not in equilibria. Depending on timescales and physics you may or may not obtain physically relevant results, but that's a whole discussion in itself.

Now, setting up spherical equilibria is currently not possible, but technically it could be implemented. From the top of my head one would need the following:
- A linearisation of the (non)-adiabatic MHD equations with desired physics, similar as done for Cartesian/cylindrical. Note that this linearisation should be performed around a 1D-varying equilibrium, so for spherical that would mean $\theta$ and $\phi$ unchanged with only variation in $r$. From a physics standpoint I think this constraint excludes some interesting cases, but maybe @jordidj can pitch in here?
- A calculation of the finite element matrix elements;
- The actual implementation of those matrix elements in the framework;
- Well-known test cases in spherical geometry to test against.

The actual implementation is quite straightforward since everything is modular and new modules can be easily added. Mathematically on the other hand it's quite a lot of work, but I guess a first step can be static, adiabatic MHD and/or with thermal effects excluded.

Keep me posted on whether or not you want to give this a shot, I'd be happy to assist in the implementation part! 

Regards,
Niels

**Reply by vreville** [reply]
Hi Niels, 

thanks for your detailed answer. OK, yes I guess I can define the equilibria in the smod_user_defined, it's essentially obtained with a Newton-Raphson method.

Yes you are right, what I am interested in is a spherical setup where the gradients of density mag field, are also along r, which may involve more complexity. In general, in the literature, this is addressed by involving spherical harmonics on the transerse direction for the perturbations. I don't know how this could be handled in Legolas, but anyway I could start with a simple hydro fully 1D case with pressure modes only. 

I'll think more about it and keep you posted.

Thanks again, 

Victor

**Reply by jordidj** [reply]
Hi Victor

As Niels pointed out, there is currently no option for a spherical equilibrium. I suppose the main reason is that in the current framework we would only have an r-dependence for a 1D equilibrium, and a Fourier exponent for the other two directions, and I don't know of many applications for which this suffices, except maybe for pulsating stars?

I think replacing the Fourier exponent for the angular directions with spherical harmonics is very non-trivial, because then the derivatives w.r.t. theta and phi are no longer equivalent to multiplying with the wave number.

If you have a clear application in mind though, I would be happy to discuss it further.

Best wishes,
Jordi
